File: frontend/src/api/approval.ts
import apiClient from './client'
import type { ApiResponse, PageResult } from './types'

export interface ApprovalItem {
  id: string
  employee_id: string
  employee_name: string
  type: string
  type_label: string
  status: string
  start_date: string | null
  end_date: string | null
  reason: string | null
  extra: Record<string, unknown> | null
  approver_id: string | null
  approver_name: string | null
  reject_reason: string | null
  created_at: string | null
}

export interface ApprovalCreateData {
  type: 'leave' | 'makeup' | 'swap' | 'expense'
  start_date?: string
  end_date?: string
  reason: string
  extra?: Record<string, unknown>
  approver_id?: string | null
}

export interface LeaveType {
  label: string
  default_days: number
}

export interface LeaveBalanceItem {
  leave_type: string
  label: string
  total_days: number
  used_days: number
  remaining_days: number
}

export interface ApproverOption {
  id: string
  name: string
  role: string
}

export interface PendingCount {
  total_pending: number
  assigned_to_me: number
}

export const approvalAPI = {
  getList(params?: { status?: string; type?: string; page?: number; page_size?: number }) {
    return apiClient.get<ApiResponse<PageResult<ApprovalItem>>>('/approvals', { params })
  },

  create(data: ApprovalCreateData) {
    return apiClient.post<ApiResponse<ApprovalItem>>('/approvals', data)
  },

  getDetail(id: string) {
    return apiClient.get<ApiResponse<ApprovalItem>>(`/approvals/${id}`)
  },

  approve(id: string) {
    return apiClient.put<ApiResponse<ApprovalItem>>(`/approvals/${id}/approve`, {})
  },

  reject(id: string, reason?: string) {
    return apiClient.put<ApiResponse<ApprovalItem>>(`/approvals/${id}/reject`, { action: 'rejected', reason: reason || null })
  },

  getLeaveTypes() {
    return apiClient.get<ApiResponse<{ types: Record<string, LeaveType> }>>('/approvals/leave-types')
  },

  getMyLeaveBalance(year?: number) {
    return apiClient.get<ApiResponse<{ year: number; balances: LeaveBalanceItem[] }>>('/approvals/leave-balance', { params: { year } })
  },

  getEmployeeLeaveBalance(employeeId: string, year?: number) {
    return apiClient.get<ApiResponse<{ year: number; balances: LeaveBalanceItem[] }>>(`/approvals/leave-balance/${employeeId}`, { params: { year } })
  },

  updateLeaveBalance(employeeId: string, leaveType: string, totalDays: number, year?: number) {
    return apiClient.put<ApiResponse<null>>(`/approvals/leave-balance/${employeeId}/${leaveType}`, null, { params: { total_days: totalDays, year } })
  },

  /** 待我审批数量（店长/老板） */
  getPendingCount() {
    return apiClient.get<ApiResponse<PendingCount>>('/approvals/pending-count')
  },

  /** 可指定的审批人列表 */
  listApprovers() {
    return apiClient.get<ApiResponse<ApproverOption[]>>('/approvals/approvers')
  },

  /** 上传报销附件图片 */
  uploadImage(file: File) {
    const formData = new FormData()
    formData.append('file', file)
    return apiClient.post<ApiResponse<{ url: string }>>('/approvals/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
  },
}

File: backend/app/schemas/approval.py
"""
统一审批 Pydantic 模型
"""
import uuid
from pydantic import BaseModel, field_validator


class ApprovalCreateRequest(BaseModel):
    type: str  # leave / makeup / swap / expense
    start_date: str | None = None
    end_date: str | None = None
    reason: str
    extra: dict | None = None
    approver_id: uuid.UUID | None = None  # 指定审批人（可选）

    @field_validator("type")
    @classmethod
    def validate_type(cls, v: str) -> str:
        if v not in ("leave", "makeup", "swap", "expense"):
            raise ValueError("审批类型必须是 leave/makeup/swap/expense")
        return v

    @field_validator("reason")
    @classmethod
    def validate_reason(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("原因不能为空")
        return v.strip()


class ApprovalReviewRequest(BaseModel):
    action: str  # approved / rejected
    reason: str | None = None  # 驳回原因

    @field_validator("action")
    @classmethod
    def validate_action(cls, v: str) -> str:
        if v not in ("approved", "rejected"):
            raise ValueError("action 必须是 approved 或 rejected")
        return v


class ApprovalItemResponse(BaseModel):
    id: uuid.UUID
    employee_id: uuid.UUID
    employee_name: str = ""
    type: str
    type_label: str = ""
    status: str
    start_date: str | None = None
    end_date: str | None = None
    reason: str | None = None
    extra: dict | None = None
    approver_id: uuid.UUID | None = None
    approver_name: str | None = None
    reject_reason: str | None = None
    created_at: str | None = None
